esmvalcore/preprocessor/_multimodel.py

```diff
@@ -248,115 +97,206 @@ def _unify_time_coordinates(cubes):
+                                     mask=True,
+                                     dtype=template_cube.data.dtype)
+
+    cube_list = []
+
+    for time_point in pad_begin:
+        new_slice = template_cube.copy()
+        new_slice.coord('time').points = float(time_point)
+        cube_list.append(new_slice)
+
+    cube_list.append(cube)
+
+    for time_point in pad_end:
+        new_slice = template_cube.copy()
+        new_slice.coord('time').points = float(time_point)
+        cube_list.append(new_slice)
+
+    cube_list = iris.cube.CubeList(cube_list)
+
+    new_cube = cube_list.concatenate_cube()
 
+    return new_cube
+
+
+def _align(cubes, span):
+    """Expand or subset cubes so they share a common time span."""
+    _unify_time_coordinates(cubes)
+
+    if _time_coords_are_aligned(cubes):
+        return cubes
+
+    all_time_arrays = [cube.coord('time').points for cube in cubes]
 
-def _assemble_data(cubes, statistic, span='overlap'):
-    """Get statistical data in iris cubes."""
-    # New time array representing the union or intersection of all cubes
-    time_spans = [cube.coord('time').points for cube in cubes]
     if span == 'overlap':
-        new_times = reduce(np.intersect1d, time_spans)
+        common_time_points = reduce(np.intersect1d, all_time_arrays)
+        new_cubes = [_subset(cube, common_time_points) for cube in cubes]
     elif span == 'full':
-        new_times = reduce(np.union1d, time_spans)
-    n_times = len(new_times)
+        all_time_points = reduce(np.union1d, all_time_arrays)
+        new_cubes = [_extend(cube, all_time_points) for cube in cubes]
+    else:
+        raise ValueError(f"Invalid argument for span: {span!r}"
+                         "Must be one of 'overlap', 'full'.")
 
-    # Target array to populate with computed statistics
-    new_shape = [n_times] + list(cubes[0].shape[1:])
-    stats_data = np.ma.zeros(new_shape, dtype=np.dtype('float32'))
+    for cube in new_cubes:
+        # Make sure bounds exist and are consistent
+        cube.coord('time').bounds = None
+        cube.coord('time').guess_bounds()
 
-    # Realize all cubes at once instead of separately for each time slice
-    _ = [cube.data for cube in cubes]
+    return new_cubes
 
-    # Make time slices and compute stats
-    for i, time in enumerate(new_times):
-        time_data = _get_time_slice(cubes, time)
-        stats_data[i] = _compute_statistic(time_data, statistic)
 
-    template = cubes[0]
-    stats_cube = _put_in_cube(template, stats_data, statistic, new_times)
-    return stats_cube
+def _combine(cubes, dim='new_dim'):
+    """Merge iris cubes into a single big cube with new dimension."""
+    equalise_attributes(cubes)  # in-place
 
+    for i, cube in enumerate(cubes):
+        try:
+            # since cubes are updated in-place, new coord may already exist
+            cube.remove_coord(dim)
+        except iris.exceptions.CoordinateNotFoundError:
+            pass
```

Comment:
```suggestion
    for i, cube in enumerate(cubes):
```

Reply:
Couldn't apply the change here, so I did it locally 👍